src/control/lib/daos/pool.go

```diff
@@ -0,0 +1,246 @@
+//
+// (C) Copyright 2020-2024 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package daos
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+
+	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
+	"github.com/daos-stack/daos/src/control/lib/ranklist"
+)
+
+type (
+	// PoolTierUsage describes usage of a single pool storage tier in
+	// a simpler format.
+	PoolTierUsage struct {
+		// TierName identifies a pool's storage tier.
+		TierName string `json:"tier_name"`
+		// Size is the total number of bytes in the pool tier.
+		Size uint64 `json:"size"`
+		// Free is the number of free bytes in the pool tier.
+		Free uint64 `json:"free"`
+		// Imbalance is the percentage imbalance of pool tier usage
+		// across all the targets.
+		Imbalance uint32 `json:"imbalance"`
+	}
+
+	// StorageUsageStats represents raw DAOS storage usage statistics.
+	StorageUsageStats struct {
+		Total     uint64           `json:"total"`
+		Free      uint64           `json:"free"`
+		Min       uint64           `json:"min"`
+		Max       uint64           `json:"max"`
+		Mean      uint64           `json:"mean"`
+		MediaType StorageMediaType `json:"media_type"`
+	}
+
+	// PoolRebuildStatus contains detailed information about the pool rebuild process.
+	PoolRebuildStatus struct {
+		Status  int32            `json:"status"`
+		State   PoolRebuildState `json:"state"`
+		Objects uint64           `json:"objects"`
+		Records uint64           `json:"records"`
+	}
+
+	// PoolInfo contains information about the pool.
+	PoolInfo struct {
+		State            PoolServiceState     `json:"state"`
+		UUID             uuid.UUID            `json:"uuid"`
+		Label            string               `json:"label,omitempty"`
+		TotalTargets     uint32               `json:"total_targets"`
+		ActiveTargets    uint32               `json:"active_targets"`
+		TotalEngines     uint32               `json:"total_engines"`
+		DisabledTargets  uint32               `json:"disabled_targets"`
+		Version          uint32               `json:"version"`
+		Leader           uint32               `json:"leader"`
+		ServiceReplicas  []ranklist.Rank      `json:"svc_reps"`
+		Rebuild          *PoolRebuildStatus   `json:"rebuild"`
+		TierStats        []*StorageUsageStats `json:"tier_stats"`
+		EnabledRanks     *ranklist.RankSet    `json:"-"`
+		DisabledRanks    *ranklist.RankSet    `json:"-"`
+		PoolLayoutVer    uint32               `json:"pool_layout_ver"`
+		UpgradeLayoutVer uint32               `json:"upgrade_layout_ver"`
+	}
+
+	// StorageTargetUsage represents DAOS target storage usage
+	StorageTargetUsage struct {
+		Total     uint64           `json:"total"`
+		Free      uint64           `json:"free"`
+		MediaType StorageMediaType `json:"media_type"`
+	}
```

Comment:
Suggestion - how about changing this to `StorageUsage` and embedding it as part of `StorageUsageStats` above? It seems like a subset of the same data, and could be genericized.

Reply:
Yeah... I think @kccain added with the `dmg pool query-targets` command. I don't see any reason not to just use the `StorageUsage` struct, though. It will have a few extra fields that aren't used in that case but I think that's fine.

Reply:
I had a similar thought last time I visited this. +1

Reply:
OK, I went ahead and moved those types too. While I was doing that, I realized that I'd forgotten to update the daos tool to use the non-dmg pretty printers, so I fixed that.